README.md
# jdatascript

Java wrapper for [DataScript](https://github.com/tonsky/datascript)

## Usage
    
[Maven](http://maven.apache.org/) dependency information:
```xml
    <dependency>
      <groupId>defng</groupId>
      <artifactId>jdatascript</artifactId>
      <version>0.5.2</version>
    </dependency>
    
    <repository>
      <id>clojars.org</id>
      <url>http://clojars.org/repo</url>
    </repository>
```

For example code, see the [tests](https://github.com/andrerichards/jdatascript/blob/master/test/defng/jdatascript/JDataScriptTest.java).

## Rationale
[DataScript](https://github.com/tonsky/datascript) has good support for JavaScript, and is usually 'advertised' as useful for 'client-side applications that need to track a lot of state', e.g. as the single store in the 'redux model'.

However, DataScript also runs on the JVM, and is generally useful for working with deeply nested data structures. 
* The datalog query engine (`q`) and `pull` API provide powerful tools to pull specific data out
* The `dbWith` API provides a convenient way of updating immutable data structures in a local scope
* The `transact` API provides a powerful way to update immutable data structures that are shared, even across multiple threads

DataScript use is convenient on the JVM from Clojure, but not so easy to use from Java, in a way that is familiar to Java programmers. JDataScript tries to fill this gap - because a tool this useful should be available to all Java programmers!

## Project status
Alpha, but there is very little code - it really just wraps the most useful DataScript functions in Java static methods.

The full DataScript API is not yet supported, but the following is provided:
* emptyDB()
* emptyDB(Map schema)
* createConn()
* createConn(Map schema)
* connFromDB(DB db)
* db(Atom conn)
* dbWith(DB db, List txData)
* transact(Atom conn, List txData)
* q(Object query, java.lang.Object... inputs)
* pull(DB db, String pattern, Object entityId)
* pullMany(DB db, String pattern, List entityIds)

Things that make it 'a bit more Java':
* enum: Operation.ADD, Operation.RETRACT
* tx(Operation operation, Object entityID, String key, Object value) : convenience method to create a transaction 


For example code, see the [tests](https://github.com/andrerichards/jdatascript/blob/master/test/defng/jdatascript/JDataScriptTest.java).

## License

Copyright © 2017 Andre Richards

Licensed under Eclipse Public License (see [LICENSE](LICENSE)).


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [0.5.2] - 2018-03-15
### Added
- pullMany : fixed return signature to be List<Map>


## [0.4.0] - 2018-03-15
### Added
- pullMany
- transact(Atom conn, Map txData) : convenience function when transacting only one tx

## [0.1.0] - 2017-11-19
### Added
- Files from the new template.
- Initial JDataScript class and tests.

[Unreleased]: https://github.com/andrerichards/jdatascript/compare/0.1.0...HEAD
[0.5.2]: https://github.com/andrerichards/jdatascript/compare/0.4.0...0.5.2
[0.4.0]: https://github.com/andrerichards/jdatascript/compare/0.1.0...0.4.0
[0.1.0]: https://github.com/andrerichards/jdatascript/compare/0.1.0...0.1.0


test/defng/jdatascript/JDataScriptTest.java
package defng.jdatascript;

import clojure.java.api.Clojure;
import clojure.lang.Atom;
import datascript.db.DB;
import defng.joculer.Joculer;
import org.junit.Test;

import java.util.List;
import java.util.Map;
import java.util.Set;

import static defng.jdatascript.JDataScript.*;
import static defng.joculer.Joculer.listOf;
import static defng.joculer.Joculer.mapOf;
import static defng.joculer.Joculer.setOf;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class JDataScriptTest {


    public static String diff(Object expected, Object actual) {
        return Joculer.diff("Expected", expected, "Actual", actual);
    }

    @Test
    public void test_dbWith() throws Exception {
        DB db = emptyDB();
        DB db1 =
                dbWith(db, listOf(listOf(":db/add", 1, "name", "Ivan"), listOf(":db/add", 1, "age", 17)));

        DB db2 = dbWith(db1, listOf(mapOf(":db/id", 2, "name", "Igor", "age", 35)));

        String query = "[:find ?n ?a :where [?e \"name\" ?n] [?e \"age\" ?a]]";

        {
            Object expected = setOf(listOf("Ivan", 17));
            Object actual = q(query, db1);
            assertTrue(diff(expected, actual), Joculer.eq(expected, actual));
        }
        {
            Object expected = setOf(listOf("Ivan", 17), listOf("Igor", 35));
            Object actual = q(query, db2);
            assertTrue(diff(expected, actual), Joculer.eq(expected, actual));
        }
    }

    @Test
    public void test_nestedMaps() throws Exception {
        String query = "[:find ?e ?a ?v :where [?e ?a ?v]]";
        DB db0 = emptyDB(mapOf("profile", mapOf(":db/valueType", ":db.type/ref")));
        {
            DB db = dbWith(db0, listOf(mapOf("name", "Igor", "profile", mapOf("email", "@2"))));
            Object actual = q(query, db);
            Object expected =
                    setOf(listOf(1, "name", "Igor"), listOf(1, "profile", 2), listOf(2, "email", "@2"));
            assertTrue(diff(expected, actual), Joculer.eq(expected, actual));
        }

        {
            DB db = dbWith(db0, listOf(mapOf("email", "@2", "_profile", mapOf("name", "Igor"))));
            Object actual = q(query, db);
            Object expected =
                    setOf(listOf(1, "email", "@2"), listOf(2, "name", "Igor"), listOf(2, "profile", 1));
            assertTrue(diff(expected, actual), Joculer.eq(expected, actual));
        }

        db0 = emptyDB(mapOf("user/profile", mapOf(":db/valueType", ":db.type/ref")));
        {
            DB db = dbWith(db0, listOf(mapOf("name", "Igor", "user/profile", mapOf("email", "@2"))));
            Object actual = q(query, db);
            Object expected =
                    setOf(listOf(1, "name", "Igor"), listOf(1, "user/profile", 2), listOf(2, "email", "@2"));
            assertTrue(diff(expected, actual), Joculer.eq(expected, actual));
        }

        {
            DB db = dbWith(db0, listOf(mapOf("email", "@2", "user/_profile", mapOf("name", "Igor"))));
            Object actual = q(query, db);
            Object expected =
                    setOf(listOf(1, "email", "@2"), listOf(2, "name", "Igor"), listOf(2, "user/profile", 1));
            assertTrue(diff(expected, actual), Joculer.eq(expected, actual));
        }
    }

    @Test
    public void test_schema() {
        Map schema = mapOf("aka", mapOf(":db/cardinality", ":db.cardinality/many"));
        DB db =
                dbWith(
                        emptyDB(schema),
                        listOf(
                                listOf(":db/add", -1, "name", "Ivan"),
                                listOf(":db/add", -1, "aka", "X"),
                                listOf(":db/add", -1, "aka", "Y"),
                                mapOf(":db/id", -2, "name", "Igor", "aka", listOf("F", "G"))));
        String query = "[:find ?aka :in $ ?e :where [?e \"aka\" ?aka]]";

        {
            Object actual = q(query, db, 1);
            Object expected = Clojure.read("#{[\"X\"] [\"Y\"]}");
            assertTrue(diff(expected, actual), Joculer.eq(expected, actual));
        }
        {
            Object actual = q(query, db, 2);
            Object expected = Clojure.read("#{[\"F\"] [\"G\"]}");
            assertTrue(diff(expected, actual), Joculer.eq(expected, actual));
        }
    }

    @Test
    public void test_pull() {
        Map schema =
                mapOf(
                        "father", mapOf(":db/valueType", ":db.type/ref"),
                        "children",
                        mapOf(":db/valueType", ":db.type/ref", ":db/cardinality", ":db.cardinality/many"));
        DB db =
                dbWith(
                        emptyDB(schema),
                        listOf(
                                mapOf(":db/id", 1, "name", "Ivan", "children", listOf(10)),
                                mapOf(":db/id", 10, "father", 1, "children", listOf(100, 101)),
                                mapOf(":db/id", 100, "father", 10)));

        {
            Map actual = pull(db, "[\"children\"]", 1);
            Map expected = mapOf("children", listOf(mapOf(":db/id", 10)));
            assertTrue(diff(expected, actual), Joculer.eq(expected, actual));
        }

        {
            List<Map> actual = pullMany(db, "[\"children\"]", listOf(1, 10));
            List expected = listOf(
                    mapOf("children", listOf(mapOf(":db/id", 10))),
                    mapOf("children", listOf(
                            mapOf(":db/id", 100),
                            mapOf(":db/id", 101))));
            assertTrue(diff(expected, actual), Joculer.eq(expected, actual));
        }

        {
            Map actual = pull(db, "[\"children\", {\"father\" [\"name\" :db/id]}]", 10);
            Map expected =
                    mapOf(
                            "children",
                            listOf(mapOf(":db/id", 100), mapOf(":db/id", 101)),
                            "father",
                            mapOf("name", "Ivan", ":db/id", 1));
            assertTrue(diff(expected, actual), Joculer.eq(expected, actual));
        }
    }

    DB peopleDB =
            dbWith(
                    emptyDB(mapOf("age", mapOf(":db/index", true))),
                    listOf(
                            mapOf(":db/id", 1, "name", "Ivan", "age", 15),
                            mapOf(":db/id", 2, "name", "Petr", "age", 37),
                            mapOf(":db/id", 3, "name", "Ivan", "age", 37)));

    @Test
    public void test_q_coll() {
        {
            Set expected = setOf(listOf(1, "Ivan"), listOf(2, "Petr"), listOf(3, "Ivan"));
            Set actual =
                    (Set)
                            q(
                                    "[:find ?e ?name :in   $ [?name ...] :where [?e \"name\" ?name]]",
                                    peopleDB,
                                    listOf("Ivan", "Petr"));
            assertTrue(diff(expected, actual), Joculer.eq(expected, actual));
        }

        {
            Set expected = setOf(listOf(1), listOf(2));
            Set actual = (Set) q("[:find ?x  :in  [?x ...] :where [(pos? ?x)]]", listOf(-2, -1, 0, 1, 2));
            assertTrue(diff(expected, actual), Joculer.eq(expected, actual));
        }
    }

    @Test
    public void test_q_relation() {
        {
            Set res =
                    (Set)
                            q(
                                    "[:find ?e ?email :in    $ $b :where [?e \"name\" ?n] [$b ?n ?email]]",
                                    peopleDB,
                                    listOf(listOf("Ivan", "ivan@mail.ru"), listOf("Petr", "petr@gmail.com")));
            Object expected =
                    setOf(listOf(1, "ivan@mail.ru"), listOf(2, "petr@gmail.com"), listOf(3, "ivan@mail.ru"));
            assertTrue(diff(expected, res), Joculer.eq(expected, res));
        }

        {
            Set res =
                    (Set)
                            q(
                                    "[:find ?e ?email  :in    $ [[?n ?email]] :where [?e \"name\" ?n]]",
                                    peopleDB,
                                    listOf(listOf("Ivan", "ivan@mail.ru"), listOf("Petr", "petr@gmail.com")));
            Object expected =
                    setOf(listOf(1, "ivan@mail.ru"), listOf(2, "petr@gmail.com"), listOf(3, "ivan@mail.ru"));
            assertTrue(diff(expected, res), Joculer.eq(expected, res));
        }
    }

    @Test
    public void test_q_rules() {
        Set res =
                (Set)
                        q(
                                "[:find ?e1 ?e2  :in  $ %  :where (mate ?e1 ?e2) [(< ?e1 ?e2)]]",
                                peopleDB,
                                "[[(mate ?e1 ?e2)   [?e1 \"name\" ?n]  [?e2 \"name\" ?n]]  [(mate ?e1 ?e2)  [?e1 \"age\" ?a]  [?e2 \"age\" ?a]]]");
        Set expected = setOf(listOf(1, 3), listOf(2, 3));
        assertTrue(diff(expected, res), Joculer.eq(expected, res));
    }

    @Test
    public void test_find_specs() {
        {
            List res = (List) q("[:find [?name ...] :where [_ \"name\" ?name]]", peopleDB);
            List expected = listOf("Ivan", "Petr");
            assertTrue(diff(expected, res), Joculer.eq(expected, res));
        }

        {
            List res =
                    (List) q("[:find [?name ?age] :where [1 \"name\" ?name] [1 \"age\" ?age]]", peopleDB);
            List expected = listOf("Ivan", 15);
            assertTrue(diff(expected, res), Joculer.eq(expected, res));
        }
        {
            String res = (String) q("[:find ?name . :where [1 \"name\" ?name]]", peopleDB);
            assertEquals("Ivan", res);
        }
    }

    @Test
    public void test_upsert() {
        Map schema = mapOf(":my/tid", mapOf(":db/unique", ":db.unique/identity"));
        Atom conn = createConn(schema);

        transact(
                conn,
                listOf(
                        mapOf(
                                ":my/tid", "5x",
                                ":my/name", "Terin")));

        transact(
                conn,
                listOf(
                        mapOf(
                                ":my/tid", "5x",
                                ":my/name", "Charlie")));

        Set names =
                (Set) q("[:find ?name :where [?e \":my/tid\" \"5x\"] [?e \":my/name\" ?name]]", db(conn));
        List expected = listOf("Charlie");
        String testOutput = diff(expected, names);
        assertTrue(testOutput, Joculer.eq(expected, names));
    }

}

src/clojure/defng/jdatascript/core.clj
(ns defng.jdatascript.core
  (:require [clojure.walk :as walk]
            [datascript.core :as d]
            [datascript.db :as db]
            [defng.joculer.core :as joculer]))

;; Conversions

(defn- keywordize [s]
  (if (and (string? s) (= (subs s 0 1) ":"))
    (keyword (subs s 1))
    s))

(defn- schema->clj [schema]
  (->>  schema
        (reduce-kv
          (fn [m k v] (assoc m k (walk/postwalk keywordize v))) {})))

(defn- pull-result->java
  [result]
  (->> result
       (walk/postwalk #(if (keyword? %) (str %) %))))

(defn- ^:declared entities->clj [entities])

(defn- entity->clj [e]
  (cond (map? e)
        (-> e
            (dissoc ":db/id")
            (assoc  :db/id (e ":db/id")))
        (= (first e) ":db.fn/call")
        (let [[_ f & args] e]
          (concat [:db.fn/call (fn [& args] (entities->clj (apply f args)))] args))
        (sequential? e)
        (let [[op & entity] e]
          (concat [(keywordize op)] entity))))

(defn- entities->clj [entities]
  (->> (joculer/->clojure entities)
       (map entity->clj)))

;; Public API

(defn db-with [db entities]
  (d/db-with db (entities->clj entities)))

(defn transact [conn entities & [tx-meta]]
  (let [entities (entities->clj entities)
        report   (-> (d/-transact! conn entities tx-meta))]
    (doseq [[_ callback] @(:listeners (meta conn))]
      (callback report))
    report))


src/java/defng/jdatascript/JDataScript.java
package defng.jdatascript;


import clojure.java.api.Clojure;
import clojure.lang.Atom;
import clojure.lang.IFn;
import datascript.db.DB;
import datascript.db.TxReport;
import defng.joculer.Joculer;

import java.util.List;
import java.util.Map;

import static defng.joculer.Joculer.listOf;

/**
 * Java wrapper functions for various DataScript functions
 */
public class JDataScript {

    static final String NAMESPACE = "defng.jdatascript.core";

    static {
        IFn require = Clojure.var("clojure.core", "require");
        require.invoke(Clojure.read("datascript.core"));
        require.invoke(Clojure.read(NAMESPACE));
    }

    static IFn GET = Clojure.var("clojure.core", "get");

    static IFn EMPTY_DB = Clojure.var("datascript.core", "empty-db");
    static IFn SCHEMA_TO_CLJ = Clojure.var(NAMESPACE, "schema->clj");
    static IFn CREATE_CONN = Clojure.var("datascript.core", "create-conn");
    static IFn CONN_FROM_DB = Clojure.var("datascript.core", "conn-from-db");
    static IFn DB = Clojure.var("datascript.core", "db");
    static IFn DB_WITH = Clojure.var(NAMESPACE, "db-with");
    static IFn TRANSACT = Clojure.var(NAMESPACE, "transact");
    static IFn Q = Clojure.var("datascript.core", "q");
    static IFn PULL = Clojure.var("datascript.core", "pull");
    static IFn PULL_MANY = Clojure.var("datascript.core", "pull-many");
    static IFn PULL_RESULT_TO_JAVA = Clojure.var(NAMESPACE, "pull-result->java");


    public static DB emptyDB() {
        return (DB) EMPTY_DB.invoke();
    }

    public static DB emptyDB(Map schema) {
        return (DB) EMPTY_DB.invoke(SCHEMA_TO_CLJ.invoke(schema));
    }
    
    public static Atom createConn() {
        return (Atom) CREATE_CONN.invoke();
    }

    public static Atom createConn(Map schema) {
        return (Atom) CREATE_CONN.invoke(schema);
    }

    public static Atom connFromDB(DB db) {
        return (Atom) CONN_FROM_DB.invoke(db);
    }

    public static DB db(Atom conn) {
        return (DB) DB.invoke(conn);
    }

    public static DB dbWith(DB db, List txData) {
        return (DB) DB_WITH.invoke(db, txData);
    }

    public static TxReport transact(Atom conn, Map txData) {
        return (TxReport) TRANSACT.invoke(conn, Joculer.toClojure(listOf(txData)));
    }

    public static TxReport transact(Atom conn, List txData) {
        return (TxReport) TRANSACT.invoke(conn, Joculer.toClojure(txData));
    }

    public static Object q(Object query, java.lang.Object... inputs) {
        Object queryVectorOrMap = query instanceof String
                ? Clojure.read((String) query)
                : query;
        if (inputs != null) {
            switch (inputs.length) {
                case 0:
                    return Q.invoke(queryVectorOrMap);
                case 1:
                    return Q.invoke(queryVectorOrMap, inputs[0]);
                case 2:
                    return Q.invoke(queryVectorOrMap, inputs[0], inputs[1]);
                case 3:
                    return Q.invoke(queryVectorOrMap, inputs[0], inputs[1], inputs[2]);
                case 4:
                    return Q.invoke(queryVectorOrMap, inputs[0], inputs[1], inputs[2], inputs[3]);
                case 5:
                    return Q.invoke(queryVectorOrMap, inputs[0], inputs[1], inputs[2], inputs[3], inputs[4]);
                case 6:
                    return Q.invoke(queryVectorOrMap, inputs[0], inputs[1], inputs[2], inputs[3], inputs[4], inputs[5]);
                case 7:
                    return Q.invoke(queryVectorOrMap, inputs[0], inputs[1], inputs[2], inputs[3], inputs[4], inputs[6]);
                case 8:
                    return Q.invoke(queryVectorOrMap, inputs[0], inputs[1], inputs[2], inputs[3], inputs[4], inputs[6], inputs[7]);
                case 9:
                    return Q.invoke(queryVectorOrMap, inputs[0], inputs[1], inputs[2], inputs[3], inputs[4], inputs[6], inputs[7], inputs[8]);
                case 10:
                    return Q.invoke(queryVectorOrMap, inputs[0], inputs[1], inputs[2], inputs[3], inputs[4], inputs[6], inputs[7], inputs[8], inputs[9]);
                case 11:
                    return Q.invoke(queryVectorOrMap, inputs[0], inputs[1], inputs[2], inputs[3], inputs[4], inputs[6], inputs[7], inputs[8], inputs[9], inputs[10]);
                case 12:
                    return Q.invoke(queryVectorOrMap, inputs[0], inputs[1], inputs[2], inputs[3], inputs[4], inputs[6], inputs[7], inputs[8], inputs[9], inputs[10], inputs[11]);
                case 13:
                    return Q.invoke(queryVectorOrMap, inputs[0], inputs[1], inputs[2], inputs[3], inputs[4], inputs[6], inputs[7], inputs[8], inputs[9], inputs[10], inputs[11], inputs[12]);
                case 14:
                    return Q.invoke(queryVectorOrMap, inputs[0], inputs[1], inputs[2], inputs[3], inputs[4], inputs[6], inputs[7], inputs[8], inputs[9], inputs[10], inputs[11], inputs[12], inputs[13]);
                case 15:
                    return Q.invoke(queryVectorOrMap, inputs[0], inputs[1], inputs[2], inputs[3], inputs[4], inputs[6], inputs[7], inputs[8], inputs[9], inputs[10], inputs[11], inputs[12], inputs[13], inputs[14]);
                case 16:
                    return Q.invoke(queryVectorOrMap, inputs[0], inputs[1], inputs[2], inputs[3], inputs[4], inputs[6], inputs[7], inputs[8], inputs[9], inputs[10], inputs[11], inputs[12], inputs[13], inputs[14], inputs[15]);
                case 17:
                    return Q.invoke(queryVectorOrMap, inputs[0], inputs[1], inputs[2], inputs[3], inputs[4], inputs[6], inputs[7], inputs[8], inputs[9], inputs[10], inputs[11], inputs[12], inputs[13], inputs[14], inputs[15], inputs[16]);
                case 18:
                    return Q.invoke(queryVectorOrMap, inputs[0], inputs[1], inputs[2], inputs[3], inputs[4], inputs[6], inputs[7], inputs[8], inputs[9], inputs[10], inputs[11], inputs[12], inputs[13], inputs[14], inputs[15], inputs[16], inputs[17]);
                case 19:
                    return Q.invoke(queryVectorOrMap, inputs[0], inputs[1], inputs[2], inputs[3], inputs[4], inputs[6], inputs[7], inputs[8], inputs[9], inputs[10], inputs[11], inputs[12], inputs[13], inputs[14], inputs[15], inputs[16], inputs[17], inputs[18]);
                case 20:
                    return Q.invoke(queryVectorOrMap, inputs[0], inputs[1], inputs[2], inputs[3], inputs[4], inputs[6], inputs[7], inputs[8], inputs[9], inputs[10], inputs[11], inputs[12], inputs[13], inputs[14], inputs[15], inputs[16], inputs[17], inputs[18], inputs[19]);
            }
        }
        return Q.invoke(queryVectorOrMap);
    }

    public static Map pull(DB db, String pattern, Object entityId) {
        return (Map) PULL_RESULT_TO_JAVA.invoke(PULL.invoke(db, Clojure.read(pattern), entityId));
    }

    public static List<Map> pullMany(DB db, String pattern, List entityIds) {
        return (List) PULL_RESULT_TO_JAVA.invoke(PULL_MANY.invoke(db, Clojure.read(pattern), entityIds));
    }

    public enum Operation {
        ADD(":db/add"),
        RETRACT(":db/retract");

        private String value;

        Operation(String value) {
            this.value = value;
        }

        public String getValue() {
            return value;
        }

        public String toString() {
            return value;
        }
    }

    public static List tx(Object entityID, String key, Object value) {
        return listOf(Operation.ADD, entityID, key, value);
    }

    public static List tx(Operation operation, Object entityID, String key, Object value) {
        return listOf(operation.value, entityID, key, value);
    }

}
